File: chapter4/quick_sort/1.rb
def quick_sort(list)
  length = list.length

  return list if length < 2

  pivot_index = (length / 2).floor
  pivot = list[pivot_index]

  new_list = list[0...pivot_index] + list[(pivot_index + 1)...length]

  lesser = []
  greater = []

  new_list.each do |x|
    if x <= pivot
      lesser.push(x)
    else
      greater.push(x)
    end
  end

  quick_sort(lesser) + [pivot] + quick_sort(greater)
end

p quick_sort([10, 5, 2, -15, 2, 3, -10, 7])

File: chapter4/quick_sort/1.exs
defmodule Grok do
  def quick_sort(list) when length(list) < 2, do: list
  def quick_sort(list) do
    list_length = length(list)

    pivot_index = floor(list_length / 2)
    pivot = Enum.at(list, pivot_index)

    new_list = List.delete_at(list, pivot_index)

    {lesser, greater} = Enum.split_with(new_list, &(&1 <= pivot))

    quick_sort(lesser) ++ [pivot] ++ quick_sort(greater)
  end
end

IO.inspect Grok.quick_sort([10, 5, 2, -15, 2, 3, -10, 7])